services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/X509CertificateCredential.java

```diff
@@ -46,74 +48,121 @@
-        super(new X500Principal(distinguishedName).getName(X500Principal.RFC2253));
+    private X509CertificateCredential(final String authId, final String issuerDN,
+            final List<X509CertificateSecret> secrets) {
+        super(authId);
         setSecrets(secrets);
+        this.issuerDN = issuerDN;
     }
 
     /**
      * Creates a new credentials object.
      * <p>
      * This method tries to decode a non-null byte array into an X.509 certificate and delegate to
-     * {@link #fromCertificate(X509Certificate)}. Otherwise, the {@link #fromSubjectDn(String, List)}
-     * method is invoked with the given distinguished name and secrets parameter values.
+     * {@link #fromCertificate(X509Certificate)}.
+     * <p>
+     * If authId, issuer DN and secretes are provided and the property
+     * {@value RegistryManagementConstants#FIELD_IS_FOR_UPDATE} is set to {@code true}, the method
+     * {@link #fromSubjectDn(String, String, List)} is invoked and the authId must be a valid subject DN.
+     * Else if the property {@value RegistryManagementConstants#FIELD_IS_FOR_UPDATE} is set to {@code false},
+     * the given authId and secrets are used to create a credentials object.
      *
      * @param derEncodedX509Certificate The DER encoding of the client certificate.
-     * @param distinguishedName The DN to use as the authentication identifier.
+     * @param authId The authentication identifier.
+     * @param isForUpdate The flag to indicate if the credentials are being decoded for an update operation or not.
+     * @param issuerDN The issuer DN of the client certificate. 
+     *                 <p> It is used to retrieve the <em>auth-id-template</em> from the tenant's
+     *                 trust anchor to generate the authentication identifier.
      * @param secrets The credential's secret(s).
      * @throws NullPointerException if certificate bytes and any of distinguished name and secrets are {@code null}.
      * @throws IllegalArgumentException if the given byte array cannot be decoded into an X.509 certificate or if
-     *                                  the given name is not a valid X.500 distinguished name or if
+     *                                  the given subject and issuer DNs are not a valid X.500 distinguished name or if
      *                                  secrets is empty.
      * @return The credentials.
      */
     @JsonCreator(mode = Mode.PROPERTIES)
     public static X509CertificateCredential fromProperties(
             @JsonProperty(value = RegistryManagementConstants.FIELD_PAYLOAD_CERT) final byte[] derEncodedX509Certificate,
-            @JsonProperty(value = RegistryManagementConstants.FIELD_AUTH_ID) final String distinguishedName,
+            @JsonProperty(value = RegistryManagementConstants.FIELD_AUTH_ID) final String authId,
+            @JsonProperty(value = RegistryManagementConstants.FIELD_IS_FOR_UPDATE) final boolean isForUpdate,
```

Comment:
IMHO we do not need this parameter. Instead, we should be able to use the presence of the issuer DN as the indicator whether we need to derive the auth-id from the subject DN or not. If an issuer DN is given (or if a certificate is being uploaded), we apply the auth-id template to the given subject DN and store it in the auth-id field.
We would then also not need to keep the issuer DN in a persistent field, would we?

Reply:
Initially I implemented as you suggested. But whenever an user updates credentials, the _auth-id_ should be always _subject-dn_ irrespective of _issuer-dn_ is provided or not. In that case, we should validate that the _auth-id_ is a valid  _subject-dn_ and format the _subject-dn_ as per RFC2253. The formatted _subject-dn_ is stored as the _auth-id_. Whenever we decode the credentials already stored in the database, we should not perform this validation as the _auth-id_ need not be always the _subject-dn_. In order to perform this, I needed that flag.

Reply:
I have strong doubts that anybody of us will be able to understand any of this in two months from now. We (I, that is) already have a hard time understanding it now ;-)
Can we maybe keep the existing properties as they are and store the generated auth-id in a separate field instead, e.g. *generated-auth-id*, which we do not expose externally but which we use to find the matching credentials when an adapter invokes *get credentials*? A Management API client would, of course, still need to supply the issuer DN if he/she wants to have a *generated-auth-id* be created as well ...

Reply:
> Can we maybe keep the existing properties as they are and store the generated auth-id in a separate field instead.


1. We can add a new database field to store the generated _auth-id_. During _get credentials_ operation we can first look into the generated auth-id field. If no value is there, we can look into the _auth-id_ field which always contain _subject-dn_
2. The _auth-id_ in the credentials object retrieved from _get credentials_ operation is compared with that of the _auth-id_ from the request during device authentication by adapters. In this case, we have to set the _generated-auth-id_ as the _auth-id_ in the credentials object.

Reply:
Step one will be done by the registry implementation, I guess.
Step two is done on the adapter, right?
> In this case, we have to set the generated-auth-id as the auth-id in the credentials object.

You mean, the registry implementation needs to return the value of the *generated-auth-id* in the returned credentials object's *auth-id* property, right? If so, then we seem to agree.

FMPOV this would make things much easier to understand because on the adapter side, nothing changes at all and we do not need to introduce additional flags and/or fields in the Registry Management API (apart from the *issuer-dn* which we would need anyway). WDYT?

Reply:
Initially I was hesitant to make any DDL changes as users need to add a new field with this Hono version. Now it seems to be a better choice as it improves the readability and helps to keep the concern separate to some extent.

Reply:
Ok, does that mean that you want to proceed into that direction?

Reply:
yes, I will implement this approach.